### PR TITLE
support decimal data type in create table

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

related #122

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by houqp — APPROVED** [review]

File: datafusion/src/sql/planner.rs [hunk 1]
```diff
@@ -372,7 +372,27 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                        return Err(DataFusionError::Internal(format!(
+                            "Error Decimal Type ({:?})",
+                            sql_type
+                        )));
+                    }
+                    (Some(p), Some(s)) => {
+                        // TODO add bound checker in some utils file or function
+                        if *p > 38 || *s > *p {
+                            return Err(DataFusionError::Internal(format!(
+                                "Error Decimal Type ({:?})",
```
Comment (Dandandan) [inline datafusion/src/sql/planner.rs:387]: A slightly more informative error would be better, such as "precision above 38  / scale bigger than precision" not supported.

File: datafusion/src/sql/planner.rs [hunk 2]
```diff
@@ -2006,6 +2026,27 @@ pub fn convert_data_type(sql: &SQLDataType) -> Result<DataType> {
...
+                    return Err(DataFusionError::Internal(format!(
+                        "Error Decimal Type ({:?})",
+                        sql_type
+                    )));
+                }
+                (Some(p), Some(s)) => {
+                    // TODO add bound checker in some utils file or function
+                    if *p > 38 || *s > *p {
+                        return Err(DataFusionError::Internal(format!(
+                            "Error Decimal Type ({:?})",
```
Comment (Dandandan) [inline datafusion/src/sql/planner.rs:2041]: Same here

File: datafusion/tests/sql.rs [hunk 3]
```diff
@@ -3761,6 +3761,28 @@ async fn register_aggregate_csv(ctx: &mut ExecutionContext) -> Result<()> {
     Ok(())
 }
 
+async fn register_simple_aggregate_csv_with_decimal_by_sql(ctx: &mut ExecutionContext) {
+    let df = ctx
+        .sql(
+            "         CREATE EXTERNAL TABLE aggregate_simple (
```
Comment (Dandandan) [inline datafusion/tests/sql.rs:3767]:
```suggestion
            "CREATE EXTERNAL TABLE aggregate_simple (
```

File: datafusion/src/sql/planner.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/src/sql/planner.rs]: Thanks for your suggestion.
It has been modified.

**Comment by liukun4515** [comment]
PTAL @Dandandan again.

**Review by Dandandan — APPROVED** [review]

**Comment by liukun4515** [comment]
@alamb please merge this.
Thanks.

**Comment by houqp** [comment]
Thanks @liukun4515

**Comment by alamb** [comment]
🎉  Thank you very much @liukun4515

File: datafusion/src/sql/planner.rs [hunk 4]
```diff
@@ -372,7 +372,27 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+            SQLDataType::Decimal(precision, scale) => {
+                match (precision, scale) {
+                    (None, _) | (_, None) => {
+                        return Err(DataFusionError::Internal(format!(
+                            "Invalid Decimal type ({:?}), precision or scale can't be empty.",
+                            sql_type
+                        )));
+                    }
+                    (Some(p), Some(s)) => {
+                        // TODO add bound checker in some utils file or function
```
Comment (alamb) [inline datafusion/src/sql/planner.rs:384]: 👍

File: datafusion/tests/sql.rs [hunk 5]
```diff
@@ -6459,3 +6481,34 @@ async fn test_select_wildcard_without_table() -> Result<()> {
     }
     Ok(())
 }
+
+#[tokio::test]
+async fn csv_query_with_decimal_by_sql() -> Result<()> {
+    let mut ctx = ExecutionContext::new();
+    register_simple_aggregate_csv_with_decimal_by_sql(&mut ctx).await;
```
Comment (alamb) [inline datafusion/tests/sql.rs:6488]: Nice

File: datafusion/src/sql/planner.rs (hunk 2 above)
Comment (alamb) [reply datafusion/src/sql/planner.rs:2041]: https://github.com/apache/arrow-datafusion/pull/1438

**Pull request merged**